src/binding/c/spawn_api.txt

```diff
@@ -0,0 +1,106 @@
+# vim: set ft=c:
+
+MPI_Comm_get_parent:
+    .desc: Return the parent communicator for this process
+    .docnotes: SignalSafe
+{
+    *parent = (MPIR_Process.comm_parent == NULL) ? MPI_COMM_NULL :
+        (MPIR_Process.comm_parent)->handle;
+}
+
+MPI_Comm_join:
+    .desc: Create a communicator by joining two processes connected by a socket
+
+MPI_Comm_spawn:
+    .desc: Spawn up to maxprocs instances of a single MPI application
+    .extra: errtest_comm_intra
+    .skip: validate-STRING_ARRAY, validate-ERROR_CODE
+{ -- error_check -- info, command, maxprocs
+    if (comm_ptr->rank == root) {
+        MPIR_ERRTEST_INFO_OR_NULL(info, mpi_errno);
+        MPIR_ERRTEST_ARGNULL(command, "command", mpi_errno);
+        MPIR_ERRTEST_ARGNEG(maxprocs, "maxprocs", mpi_errno);
+    }
+}
+{ -- handle_ptr -- info
+    MPIR_Info *info_ptr = NULL;
+    if (comm_ptr->rank == root) {
+        MPIR_Info_get_ptr(info, info_ptr);
+    }
+}
+
+MPI_Comm_spawn_multiple:
+    .desc: Spawn up multiple MPI applications
+    .skip: ThreadSafe
+    .extra: errtest_comm_intra
+    .skip: validate-STRING_ARRAY, validate-STRING_2DARRAY, validate-COMM_SIZE, validate-INFO, validate-ERROR_CODE
+{ -- error_check --
+    if (comm_ptr->rank == root) {
+        MPIR_ERRTEST_ARGNULL(array_of_info, "array_of_info", mpi_errno);
```

Comment:
in spawn_multiple, check of each  info pointer is lost
```
    if (comm_ptr->rank == root) {
        MPIR_CHKLMEM_MALLOC(array_of_info_ptrs, MPIR_Info **, count * sizeof(MPIR_Info *),
                            mpi_errno, "array of info pointers", MPL_MEM_BUFFER);
        for (i = 0; i < count; i++) {
            MPIR_Info_get_ptr(array_of_info[i], array_of_info_ptrs[i]);
        }
    }
```

Reply:
besides, when publishing the handle, the `if (intercomm_ptr) ` check seems unnecessary since a failed case will never go here
```
    mpi_errno = MPIR_Comm_spawn_multiple_impl(count, array_of_commands, array_of_argv, array_of_maxprocs, array_of_info, root, comm_ptr, &intercomm_ptr, array_of_errcodes);
    if (mpi_errno) {
        goto fn_fail;
    }
    if (intercomm_ptr) {
        MPIR_OBJ_PUBLISH_HANDLE(*intercomm, intercomm_ptr->handle);
    }
```

same for `comm_join`, `comm_spawn`

Reply:
`MPI_ERR_SPAWN` is lost in comm_spawn*

Reply:
> 
> 
> besides, when publishing the handle, the `if (intercomm_ptr) ` check seems unnecessary since a failed case will never go here
> 
> ```
>     mpi_errno = MPIR_Comm_spawn_multiple_impl(count, array_of_commands, array_of_argv, array_of_maxprocs, array_of_info, root, comm_ptr, &intercomm_ptr, array_of_errcodes);
>     if (mpi_errno) {
>         goto fn_fail;
>     }
>     if (intercomm_ptr) {
>         MPIR_OBJ_PUBLISH_HANDLE(*intercomm, intercomm_ptr->handle);
>     }
> ```
> 
> same for `comm_join`, `comm_spawn`

Probably, but I don't think it is critical. There are some places we need check for NULL pointers when returning handle objects. Always check for null pointers is simpler for the python script rather than create obscure rules for it.

Reply:
> 
> 
> `MPI_ERR_SPAWN` is lost in comm_spawn*

I don't see anywhere we return this error code.

Reply:
> 
> 
> in spawn_multiple, check of each info pointer is lost
> 
> ```
>     if (comm_ptr->rank == root) {
>         MPIR_CHKLMEM_MALLOC(array_of_info_ptrs, MPIR_Info **, count * sizeof(MPIR_Info *),
>                             mpi_errno, "array of info pointers", MPL_MEM_BUFFER);
>         for (i = 0; i < count; i++) {
>             MPIR_Info_get_ptr(array_of_info[i], array_of_info_ptrs[i]);
>         }
>     }
> ```

The allocation and populating of `array_of_info_ptrs` was moved to `spawn_impl.c` because there is no facility for the python scripts to handle info pointers array. I realize this makes `MPI_Comm_spawn` calling `MPIR_Comm_spawn_multiple` difficult since the former uses `info_ptr`. So I pushed a new commit to update python script to accept custom clean_up routine (which is needed for custom info_ptrs allocation), and have made corresponding changes in `spawn_api.txt`. Now the allocation of `array_of_info_ptrs` is in the custom handle_ptr routine inside `spawn_api.txt`